docs/docs/sample-scripts/entra/find-obsolete-m365-groups/index.mdx

```diff
@@ -0,0 +1,293 @@
+    param (
+      [Parameter(Mandatory = $false)] [Switch] $KeepOutputPath
+    )
+
+    if (!$KeepOutputPath.IsPresent -or $null -eq $Global:Path) {
+      $Script:Path = Read-Host "Set the path to the folder where you want to export the report data as csv file"
+    }
+    
+    $TestPath = Test-Path -Path  $Script:Path
+    $tStamp = (Get-Date).ToString("yyyy-MM-dd HH:mm:ss")
+    if ($TestPath -ne $true) {
+      New-Item -ItemType directory -Path $Script:Path | Out-Null
+      Write-Host "Will create file in $($Script:Path): M365GroupsReport-{current date}.csv" -ForegroundColor Yellow
+    }
+    else {
+      Write-Host "Following report file will be created in $($Script:Path): 'M365GroupsReport-$($tStamp).csv'."
+      Write-Host "`nAll data will be exported to $($Script:Path): M365GroupsReport-$($tStamp).csv." -ForegroundColor Blue
+      Write-Host "Do not edit this file during the scan." -ForegroundColor Blue
+    }
+    $Script:ReportPath = "$($Script:Path)/M365GroupsReport-$($tStamp).csv"
+  }
+
+  function Get-AllM365Groups {
+    $groups = m365 entra m365group list --includeSiteUrl | ConvertFrom-Json
+    $Script:Groups = $groups | Where-Object { $null -ne $_.siteUrl }
+  }
+
+  function Start-GroupInsightsTests {
+    Write-Host "Checking $($Script:Groups.Count) groups for activity"
+
+    $Script:Groups | ForEach-Object {
+      $groupInfo = [GroupInfo]::new()
+      $groupInfo.Reference = $_
+      $groupInfo.Membership = @{Owners = 0; Members = 0; Guests = 0 }
+      $groupInfo.TestStatus = "🟢 OK"
+
+      Write-Host "☀︎ $($groupInfo.Reference.displayName)"
+
+      # Tests
+      Test-GroupMembership -Group $groupInfo
+      Test-SharePointActivity -Group $groupInfo
+      Test-ConversationActivity -Group $groupInfo
+
+      # Report
+      New-Report -Group $groupInfo
+    }
+    
+    #Give feedback to user
+    Write-Host "`n-------------------------------------------------------------------"
+    Write-Host "`SUMMARY" -ForegroundColor DarkGreen
+    Write-Host "`-------------------------------------------------------------------"
+    Write-Host "`n👉 Found $($Global:ObsoleteGroups.Count) group$($Global:ObsoleteGroups.Count -gt 1 ? 's' : '') with possibly low activity."
+    Write-Host "`   Please review the report: " -NoNewline
+    Write-Host "$($Script:ReportPath)" -ForegroundColor DarkBlue
+  }
+
+  function Test-GroupMembership {
+    [CmdletBinding()]
+    param (
+      [Parameter(Mandatory = $true)] [GroupInfo] $Group
+    )
+
+    # Original lists
+    $users = m365 entra m365group user list --groupId $Group.Reference.id | ConvertFrom-Json
+    $owners = m365 entra m365group user list --groupId $Group.Reference.id --role Owner | ConvertFrom-Json
+    $members = m365 entra m365group user list --groupId $Group.Reference.id --role Member | ConvertFrom-Json
+    
+    # Consider guests as users that are not in the $members list
+    $guests = Compare-Object -ReferenceObject $users $members -PassThru
```

Comment:
Exactly, guests cannot be owner of a group unless they are referenced as user type `Member` in Entra ID. But the `Owner` role has no significance for the validation for the guest users anyway. 

I wanted to realize guest retrieval like this: The command `m365 entra m365group user list` retrieves all members of a group (also including guests). By adding the param `--role Member`, all guest users are filtered out. That's why I wanted to compare those two lists – and the difference between those two lists MUST be the filtered out members: the guest users.

Your suggestion to check the UPN for the string you mentioned (`#EXT#@`) is somewhat unreliable. Since the UPN of every user object in Entra ID is mutable, a system administrator could change this string in any case. Then the guest retrieval query would always return an empty result set. For that reason, I would never do a query on UPN in scripts that must have general validity. 

Do you agree, @milanholemans or am I being too pessimistic? 😃

Reply:
> I wanted to realize guest retrieval like this: The command m365 entra m365group user list retrieves all members of a group (also including guests). By adding the param --role Member, all guest users are filtered out. That's why I wanted to compare those two lists – and the difference between those two lists MUST be the filtered out members: the guest users.

That's not really true. A group consists of owners and members. When you specify `--role member`, you just list all members of the group, including guests since these are just Entra users that can be a guest from a group. The `role` parameter in this command has nothing to do with the user type.

> Your suggestion to check the UPN for the string you mentioned (#EXT#@) is somewhat unreliable. Since the UPN of every user object in Entra ID is mutable, a system administrator could change this string in any case. Then the guest retrieval query would always return an empty result set. For that reason, I would never do a query on UPN in scripts that must have general validity.

You can change it, yes, but that generally speaking not a good idea. If you want to be 100% sure, as mentioned before, we should retrieve all users and check the user type for each user to see if its a guest or not.

Reply:
Thanks for your patience and comments! 🙌 I will proceed as you told me.